### PR TITLE
#47 Reorganized options in envDynawo.sh.

### PR DESCRIPTION
**Pull request by gautierbureau**

### CODE REVIEW

File: util/envDynawo.sh [hunk 1]
```diff
@@ -21,82 +25,133 @@ error() {
...
+
+        =========== Uninstall
+        uninstall-3rd-party               uninstall all 3rd party softwares
+        uninstall-dynawo                  uninstall Dynawo
+        uninstall-all                     call uninstall-3rd-party, uninstall-dynawo
+
+        =========== Launch
+        jobs ([args])                     launch Dynawo simulation
+        jobs-with-curves ([args])         launch Dynawo simulation and open resulting curves in a browser
+        jobs-gdb ([args])                 call Dynawo's launcher with debugger with given arguments setting LD_LIBRARY_PATH correctly
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Please put a clearer description

File: util/envDynawo.sh [hunk 2]
```diff
@@ -21,82 +25,133 @@ error() {
...
+        jobs-valgrind ([args])            call Dynawo's launcher with valgrind with given arguments setting LD_LIBRARY_PATH correctly
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Same as above

Comment (rosiereflo) [reply util/envDynawo.sh]: Same for 2 next lines

File: util/envDynawo.sh [hunk 3]
```diff
@@ -1309,229 +1807,261 @@ fi
...
+    ;;
+
+  version-validation)
+    version_validation || error_exit "The current version does not fulfill the standard quality check"
     ;;
 
   *)
     echo "$1 is an invalid option"
-    echo "$usage"
+    help_dynawo
```
Comment (rosiereflo) [inline util/envDynawo.sh]: The default help should be the user one IMHO

File: util/envDynawo.sh [hunk 4]
```diff
@@ -291,46 +350,95 @@ set_environment() {
...
-  export LD_LIBRARY_PATH=$NICSLU_INSTALL_DIR/lib:$SUITESPARSE_INSTALL_DIR/lib:$SUNDIALS_INSTALL_DIR/lib64:$LIBZIP_HOME/lib:$LIBXML_HOME/lib:$LIBIIDM_HOME/lib:$DYNAWO_INSTALL_DIR/lib:$ADEPT_INSTALL_DIR/lib:$XERCESC_INSTALL_DIR/lib:$LD_LIBRARY_PATH
+ld_library_path_remove() {
+  export LD_LIBRARY_PATH=`echo -n $LD_LIBRARY_PATH | awk -v RS=: -v ORS=: '$0 != "'$1'"' | sed 's/:$//'`;
+}
 
-  if [ $LIBARCHIVE_HOME_DEFAULT != true ]; then
-    export LD_LIBRARY_PATH=$LIBARCHIVE_HOME/lib:$LD_LIBRARY_PATH
+ld_library_path_prepend() {
+  if [ ! -z "$LD_LIBRARY_PATH" ]; then
+    ld_library_path_remove $1
```
Comment (rosiereflo) [inline util/envDynawo.sh]: If I understand well we remove the dynawo environment paths and then directly put them back. Shouldn't we let the LD_LIBRARY_PATH as it is if it was already set?

File: util/envDynawo.sh [hunk 5]
```diff
@@ -291,46 +350,95 @@ set_environment() {
...
+}
 
-  if [ $BOOST_ROOT_DEFAULT != true ]; then
-    export LD_LIBRARY_PATH=$BOOST_ROOT/lib:$LD_LIBRARY_PATH
+python_path_remove() {
+  export PYTHONPATH=`echo -n $PYTHONPATH | awk -v RS=: -v ORS=: '$0 != "'$1'"' | sed 's/:$//'`;
+}
+
+python_path_append() {
+  if [ ! -z "$PYTHONPATH" ]; then
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Same as above

File: util/envDynawo.sh [hunk 6]
```diff
@@ -291,46 +350,95 @@ set_environment() {
...
+    export PYTHONPATH="$1"
+  fi
+}
+
+path_remove() {
+  export PATH=`echo -n $PATH | awk -v RS=: -v ORS=: '$0 != "'$1'"' | sed 's/:$//'`;
+}
+
+path_prepend() {
+  if [ ! -z "$PATH" ]; then
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Same as above

File: util/envDynawo.sh [hunk 7]
```diff
@@ -457,23 +565,113 @@ set_cpplint() {
   fi
 }
 
-display_environmentVariables() {
-  set -x
-  set_standardEnvironmentVariables
-  set +x
+display_environment_variables() {
+  echo ADEPT_BUILD_DIR $ADEPT_BUILD_DIR
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Very hard to maintain. Would be better to call all our environment DYNAWO_* and display all environment variables starting with that pattern. Might be done in another commit though.

File: util/envDynawo.sh [hunk 8]
```diff
@@ -518,6 +716,9 @@ build_3rd_party() {
 # Build a speficic version of third party libraries
 build_3rd_party_version() {
   if ! is_3rd_party_version_installed; then
+    if [ ! -d "$DYNAWO_SRC_DIR/3rdParty" ]; then
+      error_exit "$DYNAWO_SRC_DIR/3rdParty does not exist."
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Really useful?

File: util/envDynawo.sh [hunk 9]
```diff
@@ -564,12 +765,16 @@ uninstall_3rd_party() {
 
 # clean Dynawo
 clean_dynawo() {
-  rm -rf $DYNAWO_BUILD_DIR
+  if [ -d "$DYNAWO_BUILD_DIR" ]; then
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Useless. rm -rf does not crash if the folder doesn't exist.

File: util/envDynawo.sh [hunk 10]
```diff
@@ -564,12 +765,16 @@ uninstall_3rd_party() {
...
+    rm -rf $DYNAWO_BUILD_DIR
+  fi
 }
 
 # uninstall Dynawo
 uninstall_dynawo() {
-  rm -rf $DYNAWO_INSTALL_DIR
+  if [ -d "$DYNAWO_INSTALL_DIR" ]; then
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Same as above

File: util/envDynawo.sh [hunk 11]
```diff
@@ -679,33 +919,39 @@ build_tests() {
...
-  config_dynawo > /dev/null 2>&1 || error_exit
+  config_dynawo > /dev/null 2>&1 || error_exit "Error during config_dynawo."
   make help | grep -Ei 'tests' | grep -Eiv 'pre-tests'
   RETURN_CODE=$?
   return ${RETURN_CODE}
 }
 
+verify_browser() {
+  if [ ! -x "$(command -v $BROWSER)" ]; then
+    error_exit "Your browser $BROWSER seems not to be executable."
```
Comment (rosiereflo) [inline util/envDynawo.sh]: I would propose "Specified browser BROWSER=$BROWSER not found"

File: util/envDynawo.sh [hunk 12]
```diff
@@ -1294,6 +1587,211 @@ deploy_dynawo_autocompletion() {
...
+  if [ ! -d "$DYNAWO_HOME/documentation" ]; then
+    error_exit "$DYNAWO_HOME/documentation does not exist."
+  fi
+  cd $DYNAWO_HOME/documentation
+  bash clean.sh
+}
+
+open_doc() {
+  if [ -x "$(command -v xdg-open)" ]; then
+    xdg-open $DYNAWO_HOME/documentation/dynawoDocumentation/DynawoDocumentation.pdf
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Why not using DYNAWO_PDFVIEWER here?

Comment (rosiereflo) [reply util/envDynawo.sh]: Not sure I understand why we use first xdg-open and then the specified pdf viewer. Might be better to swap

File: util/envDynawo.sh [hunk 13]
```diff
@@ -1294,6 +1587,211 @@ deploy_dynawo_autocompletion() {
...
+  elif [ ! -z "$DYNAWO_PDFVIEWER" ]; then
+    if [ -x "$(command -v $DYNAWO_PDFVIEWER)" ]; then
+      $DYNAWO_PDFVIEWER $DYNAWO_HOME/documentation/dynawoDocumentation/DynawoDocumentation.pdf
+    else
+      error_exit "$DYNAWO_PDFVIEWER seems not to be executable."
+    fi
+  else
+    error_exit "Cannot determine how to open pdf document from command line. Use DYNAWO_PDFVIEWER environement variable."
```
Comment (rosiereflo) [inline util/envDynawo.sh]: environment

File: util/envDynawo.sh [hunk 14]
```diff
@@ -1294,6 +1587,211 @@ deploy_dynawo_autocompletion() {
...
+  for folder in $(find $DYNAWO_HOME/nrt/data -type d -name "description"); do
+    if [ -d "$folder" ]; then
+      (cd $folder; rm -f *.toc *.aux *.bbl *.blg *.log *.out *.pdf *.gz *.mtc* *.maf *.lof)
+    fi
+  done
+  (cd $DYNAWO_HOME/nrt/documentation; rm -f *.toc *.aux *.bbl *.blg *.log *.out *.pdf *.gz *.mtc* *.maf *.lof nrt_doc.tex)
+}
+
+open_nrt_doc() {
+  if [ -x "$(command -v xdg-open)" ]; then
```
Comment (rosiereflo) [inline util/envDynawo.sh]: same as above. Also we could factorize this code

File: util/envDynawo.sh [hunk 15]
```diff
@@ -1294,6 +1587,211 @@ deploy_dynawo_autocompletion() {
...
+  ld_library_path_remove $ADEPT_INSTALL_DIR/lib
+  ld_library_path_remove $XERCESC_INSTALL_DIR/lib
+  ld_library_path_remove $DYNAWO_INSTALL_DIR/lib
+  ld_library_path_remove $LIBARCHIVE_HOME/lib
+  ld_library_path_remove $BOOST_ROOT/lib
+  ld_library_path_remove $GTEST_ROOT/lib64
+  ld_library_path_remove $GTEST_ROOT/lib
+  path_remove $INSTALL_OPENMODELICA/bin
+  python_path_remove $DYNAWO_SCRIPTS_DIR
+  unset ADEPT_BUILD_DIR
```
Comment (rosiereflo) [inline util/envDynawo.sh]: Same as above: hard to maintain

File: util/envDynawo.sh (hunk 4 above)
Comment (gautierbureau) [reply util/envDynawo.sh]: The function is generic, it is to avoid prepending multiple times the same path

File: util/envDynawo.sh (hunk 7 above)
Comment (gautierbureau) [reply util/envDynawo.sh]: I agree but the previous was not satisfactory for me and not very useful. this one could be used for logs if maintained correctly. Starting all by DYNAWO is something I would have done, and it avoids settings variables that could exist otherwise either

File: util/envDynawo.sh (hunk 14 above)
Comment (gautierbureau) [reply util/envDynawo.sh]: ok

File: util/envDynawo.sh (hunk 12 above)
Comment (gautierbureau) [reply util/envDynawo.sh]: I swapped

File: util/envDynawo.sh (hunk 8 above)
Comment (gautierbureau) [reply util/envDynawo.sh]: Not really but if somehow the 3rdParty directory is deleted the cd below fails. We could also do
cd $DYNAWO_SRC_DIR/3rdParty || error_exit "$DYNAWO_SRC_DIR/3rdParty does not exist."
